flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/functions/SystemFunctionUtils.java

```diff
@@ -606,71 +606,176 @@ public static String castToString(Object object) {
+        }
+        if (object instanceof Float) {
+            return ((Float) object).intValue();
+        }
+        String stringRep = castToString(object);
+        try {
+            return Integer.valueOf(stringRep);
+        } catch (NumberFormatException e) {
+            return Double.valueOf(stringRep).intValue();
+        }
     }
 
     public static Long castToLong(Object object) {
         if (object == null) {
             return null;
         }
-        return Long.valueOf(castObjectIntoString(object));
+        if (object instanceof Boolean) {
+            return (Boolean) object ? 1L : 0L;
+        }
+        if (object instanceof BigDecimal) {
+            return ((BigDecimal) object).longValue();
+        }
+        if (object instanceof Double) {
+            return ((Double) object).longValue();
+        }
+        if (object instanceof Float) {
+            return ((Float) object).longValue();
+        }
+        String stringRep = castToString(object);
+        try {
+            return Long.valueOf(stringRep);
+        } catch (NumberFormatException e) {
+            return Double.valueOf(stringRep).longValue();
+        }
     }
 
     public static Float castToFloat(Object object) {
         if (object == null) {
             return null;
         }
+        if (object instanceof Boolean) {
+            return (Boolean) object ? 1f : 0f;
+        }
+        if (object instanceof BigDecimal) {
+            return ((BigDecimal) object).floatValue();
+        }
+        if (object instanceof Double) {
+            return ((Double) object).floatValue();
+        }
+        if (object instanceof Float) {
+            return (Float) object;
+        }
         return Float.valueOf(castObjectIntoString(object));
     }
 
     public static Double castToDouble(Object object) {
         if (object == null) {
             return null;
         }
+        if (object instanceof Boolean) {
+            return (Boolean) object ? 1d : 0d;
+        }
+        if (object instanceof BigDecimal) {
+            return ((BigDecimal) object).doubleValue();
+        }
+        if (object instanceof Double) {
+            return (Double) object;
+        }
+        if (object instanceof Float) {
+            return ((Float) object).doubleValue();
+        }
         return Double.valueOf(castObjectIntoString(object));
     }
 
     public static BigDecimal castToBigDecimal(Object object, int precision, int scale) {
         if (object == null) {
             return null;
         }
+        if (object instanceof Boolean) {
+            object = (Boolean) object ? 1 : 0;
+        }
         BigDecimal bigDecimal =
                 new BigDecimal(castObjectIntoString(object), new MathContext(precision));
-        bigDecimal = bigDecimal.setScale(scale, BigDecimal.ROUND_HALF_UP);
+        bigDecimal = bigDecimal.setScale(scale, RoundingMode.HALF_UP);
+        if (bigDecimal.precision() > precision) {
+            return null;
+        }
```

Comment:
Can the values that do not meet the requirements be warned in the form of logs, or be made known to users in other ways?

Reply:
Added a TODO comment after we have the dirty-data handling APIs in pipeline framework later.